src/hotspot/share/memory/arena.cpp

```diff
@@ -366,6 +366,10 @@ void* Arena::grow(size_t x, AllocFailType alloc_failmode) {
 // Reallocate storage in Arena.
 void *Arena::Arealloc(void* old_ptr, size_t old_size, size_t new_size, AllocFailType alloc_failmode) {
   if (new_size == 0) return NULL;
+  if (old_ptr == NULL) {
```

Comment:
If we're being consistent with realloc, probably should test this first, then for new_size == 0 (-> Afree, see above), then the rest.

Reply:
I would prefer to handle it as I wrote above.

(side note, I would love to rewrite the arena coding; either that or throw it away and reuse the existing Metaspace implementation instead for Arenas/RA. Metaspaces are Arenas too, they do the same thing, but Metaspace is a lot more evolved, and I am biased toward its code quality oc :)

Reply:
I fixed this by calling Afree in case new_size is 0. Note that Afree is a noop if the pointer is NULL.